q2_quality_control/quality_control.py

```diff
@@ -122,3 +128,115 @@ def evaluate_taxonomy(output_dir: str, expected_taxa: pd.DataFrame,
                misclassifications=None, underclassifications=None,
                composition_regression=None, score_plot=score_plot,
                mismatch_histogram=None, alignments=None)
+
+
+# decontam added code
+_WHOLE_NUM = (lambda x: x >= 0, 'non-negative')
+_PER_NUM = (lambda x: 1 >= x >= 0, 'between 0 and 1')
+_COL_STR = (lambda x: x in {'column_name', 'column_number'},
+            'sample_name or column_name or column_number')
+_DECON_METHOD_STR = (lambda x: x in {'frequency', 'prevalence', 'combined'},
+                     'freqeuncy, prevalence, combined')
+_BOOLEAN = (lambda x: type(x) is bool, 'True or False')
+# Better to choose to skip, than to implicitly ignore things that KeyError
+_SKIP = (lambda x: True, '')
+_valid_inputs = {
+    'asv_or_otu_table': _SKIP,
+    'meta_data': _SKIP,
+    'threshold': _PER_NUM,
+    'decon_method': _DECON_METHOD_STR,
+    'freq_concentration_column': _SKIP,
+    'prev_control_or_exp_sample_column': _SKIP,
+    'prev_control_sample_indicator': _SKIP,
+}
+
+
+def _check_inputs(**kwargs):
+    for param, arg in kwargs.items():
+        check_is_valid, explanation = _valid_inputs[param]
+        if not check_is_valid(arg):
+            raise ValueError('Argument to %r was %r, should be %s.'
+                             % (param, arg, explanation))
+
+
+def _decontam_identify_helper(track_fp, decon_method):
+    df = pd.read_csv(track_fp, sep='\t', index_col=0)
+    df.index.name = '#OTU ID'
+    # removes last column containing true/false information from the dataframe
+    df = df.drop(df.columns[[(len(df.columns)-1)]], axis=1)
+    if decon_method == 'combined':
+        df = df.fillna(0)
+
+    # removes all columns that are completely empty
+    temp_transposed_table = df.transpose()
+    temp_transposed_table = temp_transposed_table.dropna()
+    df = temp_transposed_table.transpose()
+    metadata = transform(df, from_type=pd.DataFrame,
```

Comment:
Since you have the transformer, you could return `pd.DataFrame` from `decontam_indentify` instead of `DecontamScoreFormat`.

I figure this kind of just evolved over time.

Reply:
As some context: the inputs and outputs will always go through the "canonical" format which is your `DecontamScoreFormat` as they become artifacts before/after any given action. From there, you inevitably have to read it, so I would lean on the transformers to make your life easier.

So there's no practical difference to the end user, but there is a slight difference (as you note) to the function itself. I would argue the dataframe is an easier API to work with if anyone was using the literal function (such as a unit-test).

Reply:
Oh okay so in the `decontam-identify` action I return the decontam score data from R as a dataframe, which I then transform into `DecontamScoreFormat` to return. Should I transform back into a dataframe after this step or just not transform the data into `DecontamScoreFormat` at all?

Reply:
Since you do some manipulation in the identify-helper, I would just skip this line (which we're commenting on) and set the return annotation of the action to pd.DataFrame (the exact same transformer will end up called in the end by the framework automatically).

Basically the R -> Python part looks good to me, and since you already have to read it as a dataframe for your own sanity, I would just leave it like that.

Reply:
Okay cool I returned a pd.Dataframe obj. In the subsequent methods I read the pd.Dataframe obj in as a qiime.metadata obj; Everything seems to be working okay, is this fine? It just fit with the existing framework better.